File: Leetcode/76.最小覆盖子串.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
public:
    string minWindow(string s, string t) {
        int min_len = INT_MAX;// 记录最小窗口的长度
        int start_idx = -1;   // 记录最小窗口的起始坐标
        int depts = t.size(); // 总欠债数
        vector<int> cnts(256, 0);

        for (char c: t) {
            cnts[c]++;
        }

        int l = 0;
        for (int r = 0; r < s.size(); r++) {
            char rc = s[r];
            if (cnts[rc] > 0) {// 说明可以还债
                depts--;
            }
            cnts[rc]--;// 不管需不需要还都可以减去这个cnt, 我们通过这个cnt去计算是否需要把欠债再加上

            while (depts == 0) {
                int current_len = r - l + 1;
                if (current_len < min_len) {
                    min_len = current_len;
                    start_idx = l;
                }

                char lc = s[l];
                if (cnts[lc] == 0) {// 这里是重点!!!!!
                    depts++;
                }
                cnts[lc]++;
                l++;
            }
        }

        return start_idx == -1 ? "" : s.substr(start_idx, min_len);
    }
};

File: Leetcode/76.最小覆盖子串 .go
package main

import "math"

func minWindow(s string, t string) string {
	cnt := map[rune]int{}
	depts := len(t)
	min_len := math.MaxInt
	start_id := -1

	for _, c := range t {
		cnt[c]++
	}

	l := 0
	for r, rc := range s {
		if cnt[rc] > 0 {
			depts--
		}
		// 减少频次, 最后会加回来所以没关系
		cnt[rc]--

		for depts == 0 {
			// 更新答案
			cur_len := r - l + 1
			if cur_len < min_len {
				min_len = cur_len
				start_id = l
			}

			var lc rune
			lc = s[l]
			// 等于零说明需要加回来, <0说明是无关字符
			if cnt[lc] == 0 {
				depts++
			}
			cnt[lc]++
			l++
		}
	}

	if start_id == -1 {
		return ""
	} else {
		return s[start_id : start_id+min_len]
	}
}
